Reorganize repo and use `pyproject.toml`

max-models opened the pull request:
Closes #24 

**Repo restructuring**

- Added a `pyproject.toml` (moved metadata from `setup.py` to `pyproject.toml`)
- Moved source files to `src/pylikwid`
- Added an `__init__.py` `src/pylikwid` which exposes all pylikwid functions to the user

**Fixes**
- All LIKWID_MAJOR, LIKWID_RELEASE, LIKWID_MINOR, VERSION, RELEASE, MINORVERSION defines passed to compiler
- Replaced `return Py_True/return Py_False` with `Py_RETURN_TRUE/Py_RETURN_FALSE`
- In `setup.py`, I added `get_include_dirs()` so `bstrlib` can be added
- In `setup.py`, I added `get_libraries()` so `os.path.join(LIKWID_INCPATH, "likwid")` is included if it exists
- In `setup.py`, I added `get_sources()` so `bstrlib.c` is added `if not (major == 5 and release >= 4):`

TomTheBear committed:
Add parsed values to TimerData for evaluation

max-models committed:
Replaced distutils.core with setuptools since distutils have been deprecated

TomTheBear commented:
Thanks a lot. There is also the issue #24 that asks for modernization of the installation method.

max-models commented:
> Thanks a lot. There is also the issue #24 that asks for modernization of the installation method.

It wasn't exactly a lot of work from my side 😅 I really like the tool so I will probably add a few more suggestions in the future! We could definitely modernize the setup a bit.

FattiMei commented:
Has anyone an idea about what went wrong with the CI? I also forked this repository to make a modern python package out of that

max-models commented:
> Has anyone an idea about what went wrong with the CI? I also forked this repository to make a modern python package out of that

I can't remember why it failed but I will take a look!

max-models committed:
Restructured the repo, use pyproject.toml and move source code to src/pylikwid

Removed setup.py

Add back setup.py

Don't install by calling setup.py explicitly in the CI

Deleted build_ext.py, moved code to setup.py

Formatted pyproject.toml with pyproject-fmt

Reset setup.py

Remove unneccesary arguments from setup()

Updated gitignore

Query version directly from the library using the exported symbols

Fix pylikwid Extension in setup.py, import everything in src/pylikwid/__init__.py

Added get_libraries()

Added Debug LIKWID libs step

Added py_limited_api=True,

Bugfix: Added extra_link_args to force the linker to pull in all symbols

Install bstrlib

Removed install

Commented out extra_link_args

bugfix: Added get_sources() to add bstrlib

include bstrlib.h if LIKWID_MAJOR == 5 && LIKWID_RELEASE >= 4

Removed the bugfix (that didn't work)

Added missing comma

Cleanup pyproject.toml

Import find_packages

Set python requirement to >=3.8

Added include src/pylikwid/pylikwid.c to MANIFEST.in

Updated extra_args

max-models committed:
Only include bstrlib.c to sources if not (major == 5 and release >= 4) in setup.py, use `Py_RETURN_TRUE/Py_RETURN_FALSE` (#3)

* Only call topology_finalize() if topo_initialized == 1

* Only include bstrlib.c to sources if not (major == 5 and release >= 4) in setup.py

* Replaced return Py_True/return Py_False with Py_RETURN_TRUE/Py_RETURN_FALSE

* Removed if (topo_initialized == 1) guard in likwid_finalizetopology

max-models committed:
Merge branch 'master' into replace-distutils-with-setuptools

max-models commented:
> Has anyone an idea about what went wrong with the CI? I also forked this repository to make a modern python package out of that

I fixed the errors and restructured the source files (moved them into `src/pylikwid`). I can't run the workflows manually here, but I ran them in my own fork, see [here](https://github.com/max-models/pylikwid/pull/2) and now they are all passing.

Sorry for taking so long to finish this. Now I think the new setup is working and can be reviewed. @TomTheBear

TomTheBear commented:
Thanks. The other PR was more an experiment from my side how to combine a `pyproject.toml` with `setup.py` for a C-backed Python module.

max-models commented:
> Thanks. The other PR was more an experiment from my side how to combine a `pyproject.toml` with `setup.py` for a C-backed Python module.

Great! I had to do some experimenting too.
